hypothesis-python/src/hypothesis/internal/filtering.py

```diff
@@ -0,0 +1,118 @@
+dataframes, it's also pretty painful to do the constructive version by hand in
+a library; so we prefer to share all the implementation effort here.
+See https://github.com/HypothesisWorks/hypothesis/issues/2701 for details.
+"""
+
+import math
+import operator
+from decimal import Decimal
+from fractions import Fraction
+from functools import partial
+from typing import Any, Callable, Mapping, Optional, Tuple, TypeVar
+
+from hypothesis.internal.compat import ceil, floor
+
+Ex = TypeVar("Ex")
+Predicate = Callable[[Ex], bool]
+
+ConstructivePredicate = Tuple[Mapping[str, Any], Optional[Predicate]]
+"""Return kwargs to the appropriate strategy, and the predicate if needed.
+
+For example::
+
+    integers().filter(lambda x: x >= 0)
+    -> {"min_value": 0"}, None
+
+    integers().filter(lambda x: x >= 0 and x % 7)
+    -> {"min_value": 0"}, lambda x: x % 7
+
+At least in principle - for now we usually return the predicate unchanged
+if needed.
+
+We have a separate get-predicate frontend for each "group" of strategies; e.g.
+for each numeric type, for strings, for bytes, for collection sizes, etc.
+"""
+
+
+def get_numeric_predicate_bounds(predicate: Predicate) -> ConstructivePredicate:
+    """Shared logic for understanding numeric bounds.
+
+    We then specialise this in the other functions below, to ensure that e.g.
+    all the values are representable in the types that we're planning to generate
+    so that the strategy validation doesn't complain.
+    """
+    if (
+        type(predicate) is partial
+        and len(predicate.args) == 1
+        and not predicate.keywords
+    ):
+        arg = predicate.args[0]
+        if (isinstance(arg, Decimal) and Decimal.is_snan(arg)) or not isinstance(
+            arg, (int, float, Fraction, Decimal)
+        ):
+            return {}, predicate
+        options = {
+            # We're talking about op(arg, x) - the reverse of our usual intuition!
+            operator.lt: {"min_value": arg, "exclude_min": True},  # lambda x: arg < x
+            operator.le: {"min_value": arg},  # lambda x: arg <= x
+            operator.eq: {"min_value": arg, "max_value": arg},  # lambda x: arg == x
+            operator.ge: {"max_value": arg},  # lambda x: arg >= x
+            operator.gt: {"max_value": arg, "exclude_max": True},  # lambda x: arg > x
+        }
+        if predicate.func in options:
+            return options[predicate.func], None
+
+    # TODO: handle lambdas by AST analysis
+
+    return {}, predicate
+
+
+def get_integer_predicate_bounds(predicate: Predicate) -> ConstructivePredicate:
+    kwargs, predicate = get_numeric_predicate_bounds(predicate)
+
+    if "min_value" in kwargs:
+        if kwargs["min_value"] == -math.inf:
+            del kwargs["min_value"]
+        elif math.isinf(kwargs["min_value"]):
+            return {}, lambda _: False
+        elif kwargs["min_value"] != int(kwargs["min_value"]):
```

Comment:
`isclose` would be a change of behaviour, and thus unacceptable in this context... it's unfortunate if someone expected `integers().filter(lambda x: x >= epsilon)` to allow `0`, but since `0 < epsilon` there's really not much else to say.

Reply:
Ah, of course. You're totally right. I wasn't thinking about this correctly